### PR TITLE
feat: foxy opportunity changes

### PR DESCRIPTION
**Pull request by reallybeard**
## Description

- Added react-table
- Modified UI for earn modals

## Notice

Before submitting a pull request, please make sure you have answered the following:

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

If applicable, please link to the github issue and put `closes #XXXX` in your comment to auto-close the issue that your PR fixes.

## Risk

High risk on this possible. Because we changed the routes work within the def modules. In addition to that we are normalizing the earn opportunity data via the `useNormalizeOpportunities` hook.

We are also displaying the data using react-table now, which should improve performance, but need to keep an eye on too many renders from the other yearn hooks.

## Testing

To test this branch, you'll need to make sure the earn opportunities still work as expected. Deposit/Withdraw from the defi page and the asset details page for that asset.

## Screenshots (if applicable)
Screen Shot 2022-03-02 at 9 24 03 AM
Screen Shot 2022-03-02 at 9 25 01 AM

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -27,6 +27,7 @@
     "@shapeshiftoss/investor-yearn": "^1.3.3",
     "@shapeshiftoss/market-service": "^1.11.1",
     "@shapeshiftoss/swapper": "^1.15.4",
+    "@types/react-table": "^7.7.9",
```
Comment (0xdef1cafe) [inline package.json:30]: move to `devDependencies` please

File: src/components/ReactTable/ReactTable.tsx [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    {
+      columns,
+      data
+    },
+    useSortBy
+  )
+  return (
+    <Table variant='clickable' {...getTableProps()}>
+      <Thead>
+        {headerGroups.map(headerGroup => (
```
Comment (0xdef1cafe) [inline src/components/ReactTable/ReactTable.tsx:23]: please check with devtools that these and the rows below don't need a `useMemo`ing

File: src/components/StakingVaults/StakingTable.tsx [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        Header: 'APY',
+        accessor: 'apy',
+        isNumeric: true,
+        display: { base: 'none', lg: 'table-cell' },
+        Cell: ({ value, row }: { value: string; row: RowProps }) => (
+          <Tag colorScheme={row.original.expired ? 'red' : 'green'}>
+            <Amount.Percent value={value} />
+          </Tag>
+        ),
+        sortType: (a: any, b: any) =>
```
Comment (0xdef1cafe) [inline src/components/StakingVaults/StakingTable.tsx:59]: any way to hint at these types better than `any`?

File: src/features/defi/helpers/normalizeOpportunity.tsx [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+      assetId: '222',
+      fiatAmount: '100',
+      cryptoAmount: '0'
+    }
+  })
+}
+
+type NoramlizeEarnOpportunitiesProps = {
+  vaultArray: SupportedYearnVault[]
+  foxyArray: any[]
```
Comment (0xdef1cafe) [inline src/features/defi/helpers/normalizeOpportunity.tsx:97]: seems like we have enough data for a type here for `foxy`

File: src/features/defi/helpers/normalizeOpportunity.tsx [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+      expired: false,
+      chain: ChainTypes.Ethereum,
+      assetId: '222',
+      fiatAmount: '100',
+      cryptoAmount: '0'
+    }
+  })
+}
+
+type NoramlizeEarnOpportunitiesProps = {
```
Comment (0xdef1cafe) [inline src/features/defi/helpers/normalizeOpportunity.tsx:95]:
```suggestion
type NormalizeEarnOpportunitiesProps = {
```

File: src/features/defi/helpers/normalizeOpportunity.tsx [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    if (vaultsWithBalances[vault.vaultAddress]) {
+      const balances = vaultsWithBalances[vault.vaultAddress]
+      cryptoAmount = balances.cryptoAmount
+      fiatAmount = balances.fiatAmount
+    }
+    const assetCAIP19 = caip19.toCAIP19({
+      chain: vault.chain,
+      network,
+      contractType,
+      tokenId: vault.tokenAddress
```
Comment (0xdef1cafe) [inline src/features/defi/helpers/normalizeOpportunity.tsx:44]: use `assetNamespace` and `assetReference` for future compatibility

File: src/types/react-table-config.d.ts [hunk 7]
```diff
@@ -0,0 +1,122 @@
+import { BoxProps } from '@chakra-ui/react'
```
Comment (0xdef1cafe) [inline src/types/react-table-config.d.ts:1]: do we need this file if we have the `@types/react-table` package?

**Review by 0xdef1cafe — CHANGES_REQUESTED** [review]

File: src/types/react-table-config.d.ts (hunk 7 above)
Comment (reallybeard) [reply src/types/react-table-config.d.ts:1]: yes their types doesn't include all the plugins, its weird this is how they recommend you do it with this version

File: src/features/defi/helpers/normalizeOpportunity.tsx (hunk 4 above)
Comment (reallybeard) [reply src/features/defi/helpers/normalizeOpportunity.tsx:97]: We don't actually know what the foxy data will look like yet, it was stubbed before.

File: src/components/ReactTable/ReactTable.tsx (hunk 2 above)
Comment (reallybeard) [reply src/components/ReactTable/ReactTable.tsx:38]: I tried wrapping these in a useMemo didn't see any performance improvement, still getting around 30ms+

File: src/components/ReactTable/ReactTable.tsx [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+import { Flex, Table, Tbody, Td, Th, Thead, Tr, useColorModeValue } from '@chakra-ui/react'
+import { useMemo } from 'react'
+import { Column, useSortBy, useTable } from 'react-table'
+
+type ReactTableProps = {
+  columns: Column<any>[]
+  data: any[]
+}
+
+export const ReactTable = ({ columns, data }: ReactTableProps) => {
```
Comment (cjthompson) [inline src/components/ReactTable/ReactTable.tsx:11]: What's the reasoning behind building our own table vs. using a library?

**Comment by cjthompson** [comment]
I'd like to see this split up into a least 2 separate PRs. I'd like the `react-table` addition to be a stand-alone PR, then the use of the new table in other places as separate PRs.

**Pull request closed without merging**